curl/src/client.cc

```diff
@@ -93,6 +93,83 @@ namespace leatherman { namespace curl {
         }
     }
 
+    static std::function<void(string const&)> make_setup_exception_thrower(request req, CURLoption curl_opt) {
+        return [req, curl_opt](string const& msg) {
+            throw http_setup_exception(req, curl_opt,  msg);
+        };
+    }
+
+    static std::function<void(string const&)> make_download_exception_thrower(request req, std::string file_path) {
+        return [req, file_path](string const& msg) {
+            throw http_download_exception(req, file_path, msg);
+        };
+    }
+
+    static std::function<void(string const&)> make_file_exception_thrower(request req, std::string file_path, std::string temp_path = "") {
+        return [req, file_path, temp_path](string const& msg) {
+            throw http_file_exception(req, file_path, temp_path, msg);
+        };
+    }
+
+    download_temp_file::download_temp_file(request const& req, std::string const& file_path, boost::optional<boost::filesystem::perms> perms) :
+        _req(req),
+        _file_path(file_path)
+    {
+        try {
+            _temp_path = fs::path(file_path).parent_path() / fs::unique_path("temp_file_%%%%-%%%%-%%%%-%%%%");
+            _fp = boost::nowide::fopen(_temp_path.string().c_str(), "wb");
+            if (!_fp) {
+                throw http_file_exception(_req, _file_path, _("Failed to open temporary file for writing"));
+            }
+            if (!perms) {
+                return;
+            }
+
+            boost::system::error_code ec;
+            fs::permissions(_temp_path.string(), *perms, ec);
+            if (ec) {
+              _external_failure_reason = _("Failed to modify permissions of temporary file");
+              _throw_external_exception = make_file_exception_thrower(_req, _file_path);
+              cleanup();
+              // Should never reach this point, as cleanup will throw the external exception.
+            }
+        } catch (fs::filesystem_error& e) {
+            throw http_file_exception(_req, _file_path, e.what());
+        }
+    }
+
+    FILE* download_temp_file::get_fp() {
+        return _fp;
+    }
+
+    void download_temp_file::cleanup() {
+         fclose(_fp);
+         boost::system::error_code ec;
+         if (!_throw_external_exception) {
+             LOG_DEBUG("download completed, now writing result to file {1}", _file_path);
+             fs::rename(_temp_path, _file_path, ec);
+             if (!ec) {
+                 return;
+             }
+             _external_failure_reason = _("Failed to move over the temporary file's downloaded contents");
+             _throw_external_exception = make_file_exception_thrower(_req, _file_path);
+         }
+         // At this point, we do have an external exception to throw.
+         fs::remove(_temp_path, ec);
```

Comment:
I'd probably be ok swallowing this exception. It shouldn't generally happen - we should be able to delete any file we create (testing verifies that we don't do anything to prevent that). If it's missing, I don't care about it.

Reply:
That makes sense, it is an edge case. I was thinking you could have a situation where the permissions of the directory containing the temp file could somehow change so that the removal might fail (or something else), + it is handled here and is not too much code. But if it is better to swallow the exception instead, should I just ignore the error code then?

And ok, I can add the logging in.

Reply:
It seems like quite a bit of extra exception handling complexity, with all the set_throw_external_exception handling.

Reply:
If I do remove the error checking of fs::remove for the temp file, I think the set_throw.. will still need to be there because the original, failure causing external exception does need to be thrown once the temp file is cleaned up, but the external_failure_reason part can be removed.

Reply:
You handle that via the destructor. My point was to make sure cleanup doesn't throw an exception - because the exceptions there are unlikely and not really actionable - so you can rely on the destructor.